### DISCUSSION TITLE
Use Tags/Keywords to group entries

### DISCUSSION THREAD

**Discussion by gyps** [emacs-citar/citar#801] [General]
I just started to use citar to organize my bibtex references.
I really like citar and it's integration into the vertico, orderless, embark, ... framework. 
To group entries into projects and topics I wanted to use tags.
In the template tags are displayed like
``` (suffix . "          ${=key= id:15}    ${=type=:12}    ${tags keywords:*}")```
which would allow to search for them and thus narrow down to groups or topics.

This all works fine if I put the keywords into the bibtex file as `keywords={...}`. 
It would somehow be nicer if the keywords could be placed in the notes file as well but somehow citar doesn't seem to find them here. Or am I'm doing something wrong?

Even better would be an option to select multiple entries in citar and have an option to add a tag/keyword to all of them in one go.

**Comment by bdarcus** [comment]
> This all works fine if I put the keywords into the bibtex file as `keywords={...}`. It would somehow be nicer if the keywords could be placed in the notes file as well but somehow citar doesn't seem to find them here. Or am I'm doing something wrong?

It might be possible with a dedicated note package like `citar-org-roam`, where one would somehow add the keywords to the notes template, though I haven't really thought about it.  I suspect we'd need to add a dedicated tag like "note-tags".

So this is a possible feature request for `citar` and/or  `citar-org-roam`.

> Even better would be an option to select multiple entries in citar and have an option to add a tag/keyword to all of them in one go.

I doubt this would be something to add to  `citar` per se. It might be a candidate for a simple custom command, which we could maybe add to the wiki.

**Reply by gyps** [reply]
I gave it a first attempt today. It might not be very elegant (I havent programmed in elisp a lot in this century) but it works (somehow). At least under the assumption, that `citar` uses only one bibliography I can add keys to an bibtex entry using this lisp function. I also started on making an interactive version (with completing read or similar) when I noticed that I cannot solve it that way because I use citar with embark.

```elisp
(defun aa/get-bibtex-keywords-for-key (key)
  (with-temp-buffer
    (insert-file-contents  (car citar-bibliography))
    (bibtex-search-entry key)
    (let ((kwds (bibtex-text-in-field "keywords")))
      (if kwds
	  (split-string   (replace-regexp-in-string "[ \t\r\n]+" " "kwds) "[:,]")
	nil))))

(defun aa/set-bibtex-keyword-for-key (key kwd)
  (with-temp-buffer
    (insert-file-contents (car citar-bibliography))
    (bibtex-search-entry key)
    (let ((kwds  (aa/get-bibtex-keywords-for-key key)))
      (if kwds
	  (let ((kwfield (bibtex-search-forward-field "keywords")))
	    (goto-char (bibtex-start-of-text-in-field kwfield))
	    (bibtex-kill-field))
	(bibtex-next-field nil))
      (bibtex-make-field `("keywords" nil
			   ,(mapconcat #'identity
				       (if (not (member kwd kwds))
					   (cons kwd kwds)
					 kwds)
				       ":")
			   nil) t))
    (bibtex-fill-entry)
    (bibtex-clean-entry)
    (write-region (point-min) (point-max) (car citar-bibliography))))
```

**Reply by gyps** [reply]
With simple interaction to delete/add single keywords it might look like this (but incompatible with embark).

```elisp
(defun aa/bibtex-edit-keywords (key)
  (with-temp-buffer
    (insert-file-contents (car citar-bibliography))
    (bibtex-search-entry key)
    (let* ((kwds  (aa/get-bibtex-keywords-for-key key))
	   (key (completing-read "enter new tag or select tag for deletion :" kwds nil nil nil nil)))
      (if kwds
	  (let ((kwfield (bibtex-search-forward-field "keywords")))
	    (goto-char (bibtex-start-of-text-in-field kwfield))
	    (bibtex-kill-field))
	(bibtex-next-field nil))
      (setq kwds (if (member key kwds)
		     (delete key kwds)
		   (cons key kwds)))
      (bibtex-make-field `("keywords" nil ,(mapconcat #'identity kwds ":") nil) t)
      (bibtex-fill-entry)
      (bibtex-clean-entry)
      (write-region (point-min) (point-max) (car citar-bibliography)))))
```

**Comment by bdarcus** [comment]
> I also started on making an interactive version (with completing read or similar) when I noticed that I cannot solve it that way because I use citar with embark.

So you can't use `citar-select-refs` to get the keys, and then prompt for the keywords, using Embark?

PS - add syntax-highlighting to the code?

**Reply by gyps** [reply]
selecting the keys using citar is straightforward but I have no idea how to prompt for keywords to be added/deleted using embark.
